packages/react-reconciler/src/__tests__/ReactSuspenseWithNoopRenderer-test.js

```diff
@@ -4203,4 +4203,60 @@ describe('ReactSuspenseWithNoopRenderer', () => {
     });
     expect(Scheduler).toHaveYielded(['Unmount Child']);
   });
+
+  it('should resume the suspending component when work loop yields immediately after catching wakeable and when wakeable pings before resuming the work loop', async () => {
+    let cachedFoo;
+    let cachedBar;
+    const resolveCbs = new Set();
+    function useFoo() {
+      if (!cachedFoo) {
+        throw new Promise(resolve =>
+          setTimeout(() => resolve((cachedFoo = 'foo')), 50),
+        );
+      }
+      return cachedFoo;
+    }
+    function useBar() {
+      if (!cachedBar) {
+        const p = {
+          then(cb) {
+            resolveCbs.add(cb);
+            // uncomment to make this test pass
+            // if (cachedBar) {
+            //   p.resolve();
+            // }
```

Comment:
this is the important bit mentioned in the point 6 [here](https://github.com/facebook/react/issues/24864#issuecomment-1178716544)

Reply:
I think this is the essence of it. The expected contract of Wakeable is that you should be able to attach a listener after the wakeable has already resolved.

Reply:
Yup! This was a silly oversight on my part in the original repro.

Reply:
Right, I agree that this somewhat breaks the standard wakeable contract. I've also realized now that in fact, this isn't **exactly** the same as Brian's situation (see [here](https://github.com/facebook/react/pull/24876/files#r917253820)). I'm unsure how to rewrite this test to repro Brian's case though because it requires precise control over async tasks/scheduler/etc and I'm not sure how to wrangle this in the test setup to replicate it 1 to 1.

I'm only concerned about a race condition in React that has manifested because of this issue - unless this is by design. The problem I've seen is that a wakeable managed to **successfully* "ping" React but it was **accidentally** ignored due to the structure of the code/logic/etc. 

As a sideways (but related) question, I wonder... if "pings" are a thing, why do you even have "suspense retry listeners"? What do they solve and why pings are not enough?